src/main/java/prefs/ConfigFile.java

```diff
@@ -0,0 +1,152 @@
+                        }
+                ).create();
+    }
+
+    /**
+     * Attempts to create the config directory if it does not exist.
+     */
+    private void createDirectoryIfNotExist() {
+        File directory = new File(configDirectory);
+        boolean directoryExists = directory.exists() && directory.isDirectory();
+        if (!directoryExists) {
+            if (directory.mkdirs()) {
+                logger.info("Config directory created: " + directory.toString());
+            } else {
+                logger.warn("Could not create config file directory");
+            }
+        }
+    }
+
+    /**
+     * Returns the File representation at the config file path
+     */
+    private File getConfigFileInDisk() {
+        return new File(configDirectory, configFileName).getAbsoluteFile();
+    }
+
+    //--- Public methods
+
+    /**
+     * Saves the config object to a file in disk
+     * @param config The config object to be saved.
+     */
+    public void saveConfig(Config config) {
+        File configFileInDisk = getConfigFileInDisk();
+        try {
+            Writer writer = new OutputStreamWriter(new FileOutputStream(configFileInDisk), CHARSET);
+            gson.toJson(config, config.getClass(), writer);
+            writer.close();
+            logger.info("Save successful: " + configFileInDisk.toString());
+
+        } catch (IOException e) {
+            HTLog.error(logger, e);
+            logger.warn("Save unsuccessful: " + configFileInDisk.toString());
+        }
+    }
+
+    /**
+     * Loads a config object from a config file in disk. The returned config object is wrapped in an Optional and may be
+     * null. A null Config is returned when either the file does not exist, or exists but deserialisation results
+     * in null.
+     * @param config The config object of the type to load
+     * @return An Optional of nullable Config.
+     */
+    public Optional<Config> loadConfig(Config config) {
+        File configFileInDisk = getConfigFileInDisk();
+        if (!configFileInDisk.exists()) {
+            logger.warn("Config file does not exist! " + configFileInDisk.toString());
+            return Optional.ofNullable(null);
+        }
+        try {
+            Reader reader = new InputStreamReader(new FileInputStream(configFileInDisk), CHARSET);
+            Optional<Config> loadedConfig = Optional.ofNullable(gson.fromJson(reader, config.getClass()));
+            reader.close();
+            logger.info("Load successful: " + configFileInDisk.toString());
+            return loadedConfig;
+
+        } catch (IOException e) {
+            HTLog.error(logger, e);
+            logger.warn("Config file could not be loaded! " + configFileInDisk.toString());
+            return Optional.ofNullable(null);
+        }
+    }
+
+    /**
+     * Attempts to delete the config file in disk.
+     * @return True if and only if the file is deleted; False otherwise.
+     */
+    public boolean delete() {
```

Comment:
Looking at this method, it seems that we only use this for testing purpose. Unless we ever would like to delete a `ConfigFile` in the disk, I don't think this method should be here. Instead, we can have such method inside the test class. Ideally, we use (urgh) reflection for `getConfigFileInDisk` to get the file. An alternative would be to delete manually from the `ConfigFile` parameter (which is more brittle if the way we craft the file path changes, e.g. instead of passing in directory and filename separately, we pass the full path).